### PR TITLE
[ConstraintSystem] Refactor storage of potential bindings

### PR DESCRIPTION
**Pull request by xedin**
- Make it possible for `PotentialBinding` to be stored in set/map

- Switch from using `SmallVector` to `SetVector` for potential binding storage in `PotentialBindings`
   This is one more step towards incremental binding computation that also simplifies the logic around
   binding storage. It's no longer necessary to pass type cache to the inference/finalization methods.

- Since constraint system is associated with potential bindings there is no need to pass it an argument
  to inference methods anymore.

### CODE REVIEW

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Comment by xedin** [comment]
@swift-ci please test source compatibility

File: include/swift/Sema/ConstraintSystem.h [hunk 1]
```diff
@@ -4623,7 +4623,10 @@ class ConstraintSystem {
   /// bindings, e.g., the supertypes of a given type.
   struct PotentialBinding {
     /// The type to which the type variable can be bound.
-    Type BindingType;
+    ///
+    /// Note that the type is mutable since it could be
+    /// adjusted as a result of type-join operation.
+    mutable Type BindingType;
```
Comment (xedin) [inline include/swift/Sema/ConstraintSystem.h:4629]: @hborla I think we can actually redo this now. Currently we-update the type but don't update the type cache. I think instead we should remove and re-introduce the binding at the same spot.

**Comment by xedin** [comment]
@swift-ci please test source compatibility

File: include/swift/Sema/ConstraintSystem.h (hunk 1 above)
Comment (xedin) [reply include/swift/Sema/ConstraintSystem.h]: Done!

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Review by hborla — APPROVED** [review]

**Comment by xedin** [comment]
I'm going to run macOS smoke test once again

**Comment by xedin** [comment]
@swift-ci please smoke test macOS platform

**Pull request merged**